Add alternative LRUCache version

lgbo-ustc opened the pull request:
Changelog category (leave one):
- Not for changelog (changelog entry is not required)


Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add evict policy support in LRUCache which wold be used in #31104


Detailed description / Documentation draft:

Add LRUCache implementation which can have unreleasable keys and manages their release state itself.

lgbo-ustc committed:
Add evict policy in LRUCache

lgbo-ustc committed:
remove unused errorcode

kssenii commented on src/Common/LRUCache.h:
```diff
@@ -6,13 +6,12 @@
 #include <atomic>
-
 #include <base/logger_useful.h>
+#include <Common/Exception.h>
```
unused header

lgbo-ustc committed:
remove unused header

kssenii commented on src/Common/LRUCache.h:
```diff
@@ -22,13 +21,32 @@ struct TrivialWeightFunction
+template <typename T>
+struct TrivialLRUCacheEvitPolicy
+{
+    // To note that the arg could be null
```
I think we can remove this comment (and the same below), because it is more like implementation detail you used.

lgbo-ustc committed:
update comments

lgbo-ustc committed:
fixe typos

kssenii commented on src/Common/LRUCache.h:
```diff
@@ -113,18 +149,39 @@ class LRUCache
-            result = true;
+            // setImpl() may fail, but the final behavior seems not be affected
+            // next call of getOrTrySet() will still call load_func()
+            is_value_updated = setImpl(key, token->value, cache_lock);
+            is_value_loaded = true;
```
@alexey-milovidov I don't know what to do with these two almost identical variables. The difference is that previously if we put key in cache it could be removed just after we did that by calling `removeOverflow()` at the end of `setImpl`. And `is_value_loaded` could  mean that it was loaded but removed from cache right after. Therefore now appears another variable (`is_value_updated`) to check that we loaded and actually managed to put it.
I would rather have one variable instead of two almost identical, but this would change previous behaviour of `getOrSet` method a bit.

kssenii replied:
Resolved, not possible to unite.

kssenii commented on src/Common/LRUCache.h:
```diff
@@ -22,13 +20,30 @@ struct TrivialWeightFunction
 };
 
+template <typename T>
+struct TrivialLRUCacheEvitPolicy
```
typo
```suggestion
struct TrivialLRUCacheEvictPolicy
```

kssenii commented on src/Common/LRUCache.h:
```diff
@@ -60,20 +75,41 @@ class LRUCache
+    ///    if is_value_loaded = false, is_value_updated = false
     template <typename LoadFunc>
-    std::pair<MappedPtr, bool> getOrSet(const Key & key, LoadFunc && load_func)
+    std::tuple<MappedPtr, bool, bool> getOrTrySet(const Key &key, LoadFunc && load_func)
```
Better to return result as struct
```
struct Result
{
    MappedPtr value;
    bool cache_miss = true;
    bool set_successful = false;
}
```

kssenii replied:
Same for `getOrSet`. But with comment for `set_successful` that for `getOrSet` it is not trustworthy because `removeOverflow` is called right after set.

lgbo-ustc replied:
Do you mean also using Result as the return of `getOrSet()`?

kssenii replied:
Yes, for consistency, because it will look strange that `getOrSet` returns `std::pair` and `getOrTrySet` returns `struct`.

kssenii replied:
@lgbo-ustc  Actually lets remove `getOrTrySet` at all. As I remember we don't really need it.

lgbo-ustc replied:
If so, maybe `getOrTrySet()` could be deleted

kssenii replied:
Yes, I just wrote it above.

lgbo-ustc replied:
done

kssenii commented on src/Common/LRUCache.h:
```diff
@@ -286,6 +344,15 @@ class LRUCache
         if (inserted)
         {
+            auto value_weight = mapped ? weight_function(*mapped) : 0;
+            // move removeOverflow() ahead here. In default, the final result is the same as the old implementation
```
I would rather remove such comments.

lgbo-ustc committed:
some modification in LRUCache

lgbo-ustc committed:
remove getOrTrySet

lgbo-ustc committed:
update comment

kssenii commented on src/Common/LRUCache.h:
```diff
@@ -113,18 +147,39 @@ class LRUCache
         /// (The token may be absent because of a concurrent reset() call).
-        bool result = false;
+        bool is_value_loaded = false;
+        bool is_value_updated = false;
```
```suggestion
        bool is_value_loaded_and_set = false;
```

kssenii commented on src/Common/LRUCache.h:
```diff
@@ -298,6 +362,13 @@ class LRUCache
         }
         else
         {
+            if (!evict_policy.canRelease(cell.value))
+            {
+                // the old value is referred by someone, cannot release now
+                // in default policy, it is always true.
+                return false;
+            }
+            evict_policy.release(cell.value); // release the old value. this action is empty in default policy.
```
```suggestion
            if (!evict_policy.canRelease(cell.value))
                return false;
            evict_policy.release(cell.value);
```

kssenii commented on src/Common/LRUCache.h:
```diff
@@ -306,17 +377,18 @@ class LRUCache
         current_size += cell.size;
 
-        removeOverflow();
+        return true;
```
Looks incorrect. In this case if there already was a value with the same key in `cells` (case `!inserted`), then when we update its value, it can have a different size (because we change previous value to `Mapped`), so there can be overflow, but now there is no `removeOverflow` at the end of this method.

lgbo-ustc replied:
updated

lgbo-ustc committed:
update variable names

lgbo-ustc committed:
update the evict action when overwrite a key

kssenii commented on src/Common/LRUCache.h:
```diff
@@ -276,16 +332,24 @@ class LRUCache
 
         if (inserted)
         {
+            if (!removeOverflow(value_weight))
+            {
+                // cannot find enough space to put in the new value
+                cells.erase(it);
+                return false;
+            }
+
             try
             {
                 cell.queue_iterator = queue.insert(queue.end(), key);
```
Also this changes previous behaviour in case in try catch below we get an exception. I mean now we could remove some values with `removeOverflow`, then get exception and not insert anything, but we already made place for this insertion. (Even though this is an exception from insertion into list and not in general likely to happen). So better to first insert into queue, remember iterator to currently inserted key, then check if there is enough space, it not -- remove from queue.

lgbo-ustc replied:
done

kssenii replied:
It was a bad idea actually. Or now we need to guarantee that for key, which we want to set, `canRelease` must return true, and also one more problem -- `release()` (is it legal to call it if in fact we did not return from `set`?). For cache it removes files from local fs, and BTW does not currently check if path exists before doing that.

I'll think what to do with that, how to do it differently. If you also have ideas -- propose.
Also will need to add unit tests for changes from this pr.

kssenii replied:
Ok no problem here after all because value will be `nullptr`, but then need to move check for nullptr out of `release()` and `canRelease()` and pass a reference inside.

lgbo-ustc replied:
btw, I don't find where to add a unit test

kssenii replied:
Similar to `src/Common/tests/gtest_lru_hash_map.cpp`
to run:
```
 ninja unit_tests_dbms
 cd build/src
./unit_tests_dbms --gtest_filter='TransformQueryForExternalDatabase*' --gtest_brief=0 --gtest_throw_on_failure=1 
```

lgbo-ustc replied:
Another fatal bug, when `canRelease()` returns true, before calling `release()`, the element may turn into unreleasable again

kssenii replied:
> Another fatal bug, when canRelease() returns true, before calling release(), the element may turn into unreleasable again

As I understood you, what you say should not happen because we hold mutex in `set` and in `get`. It may become unreleasable if someone successfully called `get`, but while we are in `trySet` this will not happen.

lgbo-ustc replied:
following is the case
1. we get A from cache, A is releasable
2. then we try to set B and sees A is releasable, 
3. at now something happpens on A outside lru cache, and make A unreleasable
4. then lru cache call release A to make space for B.

kssenii replied:
This is the case I asked you to fix here https://github.com/ClickHouse/ClickHouse/pull/31104#discussion_r775749309.
So now I believe that this is case is handled by your code of cache for storage because getting key from cache and creating a value holder object is done under mutex. And getting/setting keys in cache always goes throw that mutex.

lgbo-ustc replied:
In our use case, there is and outside mutex which could avoid this case, but we cannot gurantee for other use cases

kssenii replied:
Ok I agree, then may be instead of `bool canRelease` and `void release` have just `bool tryRelease()`?

lgbo-ustc replied:
I just thought use `markRelease` for mark the element could be used anymore and `unmarkRelease` for rollback `removeOverflow`, but this may still have side effect

kssenii replied:
When would we need to rollback it?

lgbo-ustc replied:
It seems more likely what we need is a resource pool with key index and lru evict policy. The apis may are as following
```
// if could not insert into the cache, return nullptr
// if result is not null,  increase its reference
// use lru evict policy inside to evict elements which reference is 0
std::shared_ptr<Value> acquire(const Key & key,  LoadFunc && load_func);
// decrease its reference
void release(const Key & key, std::shared_ptr<Value> v);
```

lgbo-ustc replied:
> When would we need to rollback it? May be just make implementation when we would not need rollback?

for example, we have elements : a,b,c, and a is releasabel, other two not. Now we want to insert d, but we cannot find enough space for d even after evict a. Maybe we just want to keep the cache elements as before. a, b and c are still in the cache

kssenii replied:
> I still prefer a new implementation but not LRUCache

Yes, let's try to make a new one. I also start to think it will be better.

lgbo-ustc committed:
optimization

lgbo-ustc committed:
update setImpl()

kssenii commented on src/Common/LRUCache.h:
```diff
@@ -113,18 +147,39 @@ class LRUCache
         {
-            setImpl(key, token->value, cache_lock);
-            result = true;
+            // setImpl() may fail, but the final behavior seems not be affected
+            // next call of getOrTrySet() will still call load_func()
+            is_value_loaded_and_set = setImpl(key, token->value, cache_lock);
```
```suggestion
            is_value_loaded_and_set = setImpl(key, token->value, cache_lock);
```

lgbo-ustc committed:
remove comments

kssenii commented on src/Common/LRUCache.h:
```diff
@@ -295,28 +350,42 @@ class LRUCache
+                return false;
+            if (value_weight > cell.size && !removeOverflow(value_weight - cell.size))
+                return false;
+            evict_policy.release(cell.value); // release the old value. this action is empty in default policy.
```
Also I am worried that `release` can throw exception...
But not sure should we catch it here or just make sure in `release` method that it never throws.

kssenii replied:
Ok I the only solution is to make `release` and `canRelease` as `noexcept`. Because it is controversial if or how we should process exceptions from `release` for old keys.

lgbo-ustc committed:
Add LRUResourceCache

1. add LRUResourceCache for managing resource cache in lru policy
2. rollback LRUCache to the original version
3. add remove() in LRUCache
4. add unit tests for LRUResourceCache and LRUCache

lgbo-ustc commented:
@kssenii plz see the new cache, `LRUResourceCache`

lgbo-ustc committed:
fix a bug

lgbo-ustc committed:
add a new test case

lgbo-ustc committed:
add a new test case

lgbo-ustc committed:
update destructor

lgbo-ustc committed:
fixed code style

kssenii commented on src/Common/LRUResourceCache.h:
```diff
@@ -0,0 +1,255 @@
+    }
+
+    // key mustn't be in the cache
+    Cell * insert_value(const Key & insert_key, MappedPtr value)
```
Style. Should be `insertValue` at least. But I like name `set(...)` more (or `setImpl`).

kssenii commented on src/Common/tests/gtest_lru_resource_cache.cpp:
```diff
@@ -0,0 +1,276 @@
+
+    succ = mcache.tryRemove(1);
+    ASSERT_TRUE(!succ);
+    val = mcache.acquire(1);
```
Ok but I also stared to think that it makes sense to have a method (not sure how to name it, may be just `get`). It will return a holder, which will call `acquire` in constructor and `release` in desctructor. And may be make public only this method, `acquire` and `release` may be private. What do you think?

lgbo-ustc replied:
good idea, I take a try

lgbo-ustc replied:
updated

lgbo-ustc committed:
add MappedHolder to get cache values

lgbo-ustc committed:
format code style

kssenii commented on src/Common/LRUResourceCache.h:
```diff
@@ -0,0 +1,303 @@
+        {
+            return *(val.get());
+        }
+        static bool tryRemove(std::unique_ptr<MappedHolder> * holder_ptr)
```
Why would we need such method? It seems redundant to me.

lgbo-ustc replied:
In our case, when a remote file is updated, we need to drop the current cached file

kssenii replied:
Ok, but then may be it makes sense to make it's implementation such that if refcount is 0, it removes the key, if non-zero refcount, then it sets the flag in cell `expired = true`. 
So if `expired == true`, then 
1. `get` for this key will return false
2. the weight of value if this key is still taken into account when we check remaining space
3. when `release` is called for this key and refcount gets to 0, then we remove this key from cache

Also I see no reason why it should accept a holder, and not just key. If we have a holder and then decide that corresponding key needs to be removed, then better to get rid of holder first. I mean no need in such method in holder to me.

lgbo-ustc replied:
done

kssenii commented on src/Common/LRUResourceCache.h:
```diff
@@ -0,0 +1,295 @@
+    {
+    public:
+        ~MappedHolder() { cache->release(key); }
+        Mapped & value() { return *(val.get()); }
```
```suggestion
        Mapped & value() { return *value; }
```

kssenii commented on src/Common/LRUResourceCache.h:
```diff
@@ -0,0 +1,295 @@
+    // use get() or getOrSet() to access the elements
+    MappedHolderPtr get(const Key & key)
+    {
+        auto mappedptr = getImpl(key);
```
```suggestion
        auto mapped_ptr = getImpl(key);
```

lgbo-ustc committed:
some change

lgbo-ustc committed:
update comments

lgbo-ustc committed:
rename test cases

lgbo-ustc committed:
rename test cases

kssenii commented:
@lgbo-ustc can you please enable edits by maintainers (on the right menu of this PR)? So I can also push to your branch.

lgbo-ustc commented:
![image](https://user-images.githubusercontent.com/2599447/147736713-7ed7a749-bacd-45ca-9cd2-30a4f62d32cf.png)

There is no menu here. is it enabled when this pr is crearted?

kssenii commented:
> There is no menu here. is it enabled when this pr is crearted?

hm, Ok, sad. Usually the button to allow edits just below `notifications` section. And usually when you create a pr there is also such botton, but it is true by default, i.e. usually I can push to forks without even asking to allow that manually. 
(though you are not the first who does not have this allow edits button)

kssenii commented on src/Common/LRUResourceCache.h:
```diff
@@ -0,0 +1,307 @@
+    template <typename LoadFunc>
+    MappedPtr getImpl(const Key & key, LoadFunc && load_func)
+    {
+        InsertToken * insert_token = nullptr;
```
Why don't you add `InsertTokenHolder`?

lgbo-ustc replied:
updated

lgbo-ustc committed:
add token holder

lgbo-ustc commented on src/Common/LRUResourceCache.h:
```diff
@@ -0,0 +1,307 @@
+    template <typename LoadFunc>
+    MappedHolderPtr getOrSet(const Key & key, LoadFunc && load_func)
+    {
+        auto mappedptr = getImpl(key, load_func);
```
done

kssenii commented on src/Common/LRUCache.h:
```diff
@@ -64,6 +64,17 @@ class LRUCache
+            return;
+        auto & cell = it->second;
+        current_size -= cell.size;
+        cells.erase(it);
```
Forgot to also remove from `queue`?
btw let's remove it since it is not needed now?

lgbo-ustc replied:
fixed

kssenii commented on src/Common/tests/gtest_lru_resource_cache.cpp:
```diff
@@ -0,0 +1,270 @@
+    int x = 2;
+    auto load_int = [&] { return std::make_shared<int>(x); };
+    auto holder1 = mcache.getOrSet(1, load_int);
+    holder1 = nullptr;
```
Can also be like this
```suggestion
    holder1.reset();
```
(not necessary)

lgbo-ustc replied:
done

lgbo-ustc committed:
fix a bug in LRUCache::remove()

kssenii reviewed (approved):
Ok, let's wait for tests and merge.
I also need this implementation for something else so possibly it will be reworked in some way (and I will test it more).

lgbo-ustc committed:
fix a bug in LRUResourceCache::getImpl()
